app/coffee/ChannelManager.coffee

```diff
@@ -0,0 +1,44 @@
+logger = require 'logger-sharelatex'
+metrics = require "metrics-sharelatex"
+settings = require "settings-sharelatex"
+
+ClientMap = new Map() # for each redis client, stores a Set of subscribed channels
+
+# Manage redis pubsub subscriptions for individual projects and docs, ensuring
+# that we never subscribe to a channel multiple times. The socket.io side is
+# handled by RoomManager.
+
+module.exports = ChannelManager =
+    _createNewClientEntry: (rclient) ->
+        ClientMap.set(rclient, new Set()).get(rclient)
+
+    subscribe: (rclient, baseChannel, id) ->
+        existingChannelSet = ClientMap.get(rclient) || @_createNewClientEntry(rclient)
+        channel = "#{baseChannel}:#{id}"
+        if existingChannelSet.has(channel)
+            logger.error {channel}, "already subscribed"
+        else
+            rclient.subscribe channel
```

Comment:
Possible race condition here. Consider adding the channel to the `existingChannelSet` first.

Reply:
Thanks, I forgot the `subscribe` is asynchronous.